# Q0: import dash standard libraries

In [1]:
from dash import Dash, dcc, html, Input, Output
import plotly.graph_objects as go
from jupyter_dash import JupyterDash
from dash import Dash, html
import pandas as pd
import plotly.express as px
import numpy as np

# Q1: Make dash that prints "Data Science"

In [ ]:
app=Dash(__name__)

app.layout=html.Div([html.H1("Data Science")])

if __name__ =="__main__":
    app.run_server()

Dash is running on http://127.0.0.1:8050/

 * Serving Flask app "__main__" (lazy loading)
 * Environment: production
   Use a production WSGI server instead.
 * Debug mode: off


 * Running on http://127.0.0.1:8050/ (Press CTRL+C to quit)
127.0.0.1 - - [31/May/2023 19:31:07] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:31:08] "GET /_dash-layout HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:31:08] "GET /_dash-dependencies HTTP/1.1" 200 -


# Q2: Make dash that takes name input and outpt "Welcome to Data Science dashboard {name}"

In [2]:
app = JupyterDash(__name__)

app.layout = html.Div([
    html.H6("Enter Your Name : "),
    html.Div([
        "Input: ",
        dcc.Input(id='my-input', value='initial value', type='text')
    ]),
    html.Br(),
    html.Div(id='my-output'),

])


@app.callback(
    Output(component_id='my-output', component_property='children'),
    Input(component_id='my-input', component_property='value'),
)
def update_output_div(input_value):
    return f'Welcome to Data Science dashboard {input_value}'


if __name__ == '__main__':
    app.run_server(mode='inline')

Dash is running on http://127.0.0.1:8050/



# Q3: Make Plotly-Dash that takes two inputs (numbers) and output 3 numbers x, y , z= x+y
- hint: use html.Div, dcc.Input(type='number')

In [2]:
# from dash import Dash, dcc, html, Input, Output
# from jupyter_dash import JupyterDash

app = JupyterDash(__name__)

app.layout = html.Div([
    html.H6("Please Enter 2 numbers : "),
    html.Div([
        "Input: ",
        dcc.Input(id='my-input', value='initial value', type='number')
    ]),
    html.Div([
        "Input: ",
        dcc.Input(id='my-input2', value='initial value 2', type='number')
    ]),
    html.Br(),
    html.Div(id='my-output'),

])


@app.callback(
    Output(component_id='my-output', component_property='children'),
    Input(component_id='my-input', component_property='value'),
    Input(component_id='my-input2', component_property='value')
)
def update_output_div(input_value, input_value2):
    return f'=: {input_value+input_value2}'


if __name__ == '__main__':
    app.run_server()


Dash is running on http://127.0.0.1:8050/

Dash app running on http://127.0.0.1:8050/


# Q4.1: Make JubyterDash with Rangeslider from 1-15 with 6 option

In [2]:
app = JupyterDash(__name__) 
# app = Dash(__name__)


app.layout = html.Div([
    html.H4('Interactive plot with custom data source'),
    
    html.P("q4.1"),
    dcc.RangeSlider(id="slider", min=1, max=15,  step=3),
])


@app.callback(
    Input("slider", "value"))
    
def Display(sl):
    return sl

# app.run()  
app.run_server(use_reloader=False, mode='inline') # Turn off reloader if inside Jupyter

Dash is running on http://127.0.0.1:8050/



# Q4.2: add to the previous dash: Label, set default value on rangeslider as 7-10

In [2]:
app = JupyterDash(__name__) 
# app = Dash(__name__)


app.layout = html.Div([
    
    
    html.P("Q4.2"),
    dcc.RangeSlider(id="slider", min=1, max=15, value=[7,10], step=3),
])


@app.callback(
    Input("slider", "value"))
    
def Display(sl):
    return sl

# app.run()  
app.run_server(use_reloader=False, mode='inline') # Turn off reloader if inside Jupyter

Dash is running on http://127.0.0.1:8050/



# Q5: Make 3 different Dcc types (Dropdown - Checkbox...) in one dash

In [2]:

app = JupyterDash(__name__)


app.layout = html.Div([
    html.H4('Q5'),
    html.P("Dropdown:"),
    dcc.Dropdown(
        id="dropdown",
        options=['test1', 'test2', 'test3'],
    ),
    html.P("CheckList"),
        dcc.Checklist(
        ['t1', 't2', 't3'],
        inline=True
    ),
    html.P("RangeSlider : "),
    dcc.RangeSlider(id="slider", min=1, max=15, value=[7,10], step=3)
])




app.run_server(mode='inline')

Dash is running on http://127.0.0.1:8050/



# Q6: Make Plotly chart, then add it to a plotly_dash

In [7]:
import plotly.express as px
import numpy as np

In [2]:
from dash import Dash, dcc, html, Input, Output
import plotly.graph_objects as go
import numpy as np
from jupyter_dash import JupyterDash


app = JupyterDash(__name__) 
# app = Dash(__name__)


app.layout = html.Div([
    html.H4('Q6'),
    dcc.Graph(id="graph"),
    html.P("Number of bars:"),
    dcc.Slider(id="slider", min=2, max=10, value=4, step=1),
])


@app.callback(
    Output("graph", "figure"), 
    Input("slider", "value"))
    
def update_bar_chart(size):
    data = np.random.uniform(3, 2, size=size)
    fig = go.Figure(
        data=[go.Bar(y=data)]
    )
    return fig

# app.run()  
app.run_server(use_reloader=False, mode='inline') 

Dash is running on http://127.0.0.1:8050/



# Q7: From our data make plotly-dash with suitble chart, and Dropdown.

- Dropdown should contain at least 3 options, each one represents different column.
- Add label to dropdown.
- The chart data should change depending on the selected dropdown option.

In [ ]:
from dash import Dash, dcc, html, Input, Output
import plotly.express as px
from dash import Dash, dcc, html, Input, Output
import plotly.express as px


import pandas as pd

df=pd.read_csv("titanic passenger list.csv")


app = Dash(__name__)

app.layout = html.Div([
    html.H4('Show Survived based on 2 x-axis '),
    dcc.Dropdown(
        id="dropdown",
        options=['embarked','sex'],
        #value=['embarked', 'sex'],
        multi=True
    ),
    dcc.Graph(id="graph"),
])


@app.callback(
    Output("graph", "figure"), 
    Input("dropdown", "value"))
def Display(x_axis):
    return px.box(data_frame=df,x=x_axis,y=df["survived"])


app.run_server()

Dash is running on http://127.0.0.1:8050/

 * Serving Flask app "__main__" (lazy loading)
 * Environment: production
   Use a production WSGI server instead.
 * Debug mode: off


 * Running on http://127.0.0.1:8050/ (Press CTRL+C to quit)
127.0.0.1 - - [31/May/2023 19:39:12] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:39:12] "GET /_dash-layout HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:39:12] "GET /_dash-dependencies HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:39:12] "GET /_dash-component-suites/dash/dcc/async-dropdown.js HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:39:12] "GET /_dash-component-suites/dash/dcc/async-graph.js HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:39:12] "GET /_dash-component-suites/dash/dcc/async-plotlyjs.js HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:39:13] "POST /_dash-update-component HTTP/1.1" 200 -


In [4]:
# from dash import Dash, dcc, html, Input, Output
# import plotly.express as px
# from dash import Dash, dcc, html, Input, Output
# import plotly.express as px
# import pandas as pd
# from dash import Dash, dcc, html, Input, Output
# import plotly.graph_objects as go
# from jupyter_dash import JupyterDash
# from dash import Dash, html

# Q8: add 2 more charts to the prevoius Ploty-Dash

In [ ]:
from dash import Dash, dcc, html, Input, Output
import plotly.express as px
from dash import Dash, dcc, html, Input, Output
import plotly.express as px


import pandas as pd

df=pd.read_csv("titanic passenger list.csv")


app = Dash(__name__)

app.layout = html.Div([
    html.H4('Show Survived based on 2 x-axis '),
    dcc.Dropdown(
        id="dropdown",
        options=['embarked','sex'],
        #value=['embarked', 'sex'],
        multi=True
    ),
    dcc.Graph(id="graph"),
    dcc.Graph(id="New_plot",figure=px.bar(data_frame=df,y=df["survived"],x=df["age"])),
    dcc.Graph(id="Pie_plot",figure=px.pie(df,df["pclass"]))
])


@app.callback(
    Output("graph", "figure"), 
    Input("dropdown", "value"))
def Display(x_axis):
    return px.box(data_frame=df,x=x_axis,y=df["survived"])


app.run_server()

Dash is running on http://127.0.0.1:8050/

 * Serving Flask app "__main__" (lazy loading)
 * Environment: production
   Use a production WSGI server instead.
 * Debug mode: off


 * Running on http://127.0.0.1:8050/ (Press CTRL+C to quit)
127.0.0.1 - - [31/May/2023 19:40:04] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:40:04] "GET /_dash-layout HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:40:04] "GET /_dash-dependencies HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:40:04] "GET /_dash-component-suites/dash/dcc/async-graph.js HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:40:04] "GET /_dash-component-suites/dash/dcc/async-dropdown.js HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:40:04] "GET /_dash-component-suites/dash/dcc/async-plotlyjs.js HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:40:04] "POST /_dash-update-component HTTP/1.1" 200 -


# Q9: Search for "Plotly Gallery" and fetch for a graph and use it with our data.

- Note: You need to put the resourse addres of fetched your grapth.

In [ ]:

# resource https://plotly.com/python/histograms/

df=pd.read_csv("titanic passenger list.csv")


app = Dash(__name__)

app.layout = html.Div([
    html.H4('Show Survived based on 2 x-axis '),
    dcc.Dropdown(
        id="dropdown",
        options=['embarked','sex'],
        #value=['embarked', 'sex'],
        multi=True
    ),
    dcc.Graph(id="graph"),
])


@app.callback(
    Output("graph", "figure"), 
    Input("dropdown", "value"))
def update_bar_chart(x_axis):
    return px.histogram(df, x=x_axis)


app.run_server()

Dash is running on http://127.0.0.1:8050/

 * Serving Flask app "__main__" (lazy loading)
 * Environment: production
   Use a production WSGI server instead.
 * Debug mode: off


 * Running on http://127.0.0.1:8050/ (Press CTRL+C to quit)
127.0.0.1 - - [31/May/2023 19:40:34] "GET / HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:40:34] "GET /_dash-layout HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:40:34] "GET /_dash-dependencies HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:40:34] "GET /_dash-component-suites/dash/dcc/async-dropdown.js HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:40:34] "GET /_dash-component-suites/dash/dcc/async-graph.js HTTP/1.1" 200 -
127.0.0.1 - - [31/May/2023 19:40:34] "GET /_dash-component-suites/dash/dcc/async-plotlyjs.js HTTP/1.1" 200 -
[2023-05-31 19:40:34,676] ERROR in app: Exception on /_dash-update-component [POST]
Traceback (most recent call last):
  File "C:\Users\ttsul\anaconda3\lib\site-packages\flask\app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "C:\Users\ttsul\anaconda3\lib\site-packages\flask\app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "C:\Us

# Challenge 1.1: Create Plotly dash with two charts depends on different columns of the data
- First Chart : Pie
- Secound Chart: is by your Choice ( Not Pie )

# Challenge 1.2: Make Hovering over a slice of the pie changes the data for the second chart depending on what the slice is.